metrics/internal/stackdriver/secret_watcher_test.go

```diff
@@ -0,0 +1,569 @@
+		return kubeclientForTest, nil
+	}
+
+	// Always reset secret
+	dsft = newSecretForTest(defaultSecretName, defaultSecretNamespace)
+}
+
+// testSetup sets up tests.
+func testSetup() {
+	testSetupWithCustomKubeclient(fclient.NewSimpleClientset())
+}
+
+// testCleanup cleans up tests.
+func testCleanup() {
+	createStackdriverKubeclientFunc = createKubeclient
+}
+
+func TestNewSecretWatcher(t *testing.T) {
+	for _, test := range defaultWatchersToTest {
+		t.Run(test.name, func(t *testing.T) {
+			testSetup()
+			test.constructor(t, defaultObserverFuncs)
+			testCleanup()
+		})
+	}
+}
+
+func TestInvalidNewSecretWatcher(t *testing.T) {
+	invalidCases := []struct {
+		name        string
+		constructor func() (SecretWatcher, error)
+	}{
+		{
+			name:        "SingleNamespaceNoNamespace",
+			constructor: func() (SecretWatcher, error) { return NewSecretWatcherSingleNamespace("") },
+		}, {
+			name:        "SingleSecretNoNamespace",
+			constructor: func() (SecretWatcher, error) { return NewSecretWatcherSingleSecret("", dsft.name) },
+		}, {
+			name:        "SingleSecretNoName",
+			constructor: func() (SecretWatcher, error) { return NewSecretWatcherSingleSecret(dsft.namespace, "") },
+		}, {
+			name:        "SingleSecretNothing",
+			constructor: func() (SecretWatcher, error) { return NewSecretWatcherSingleSecret("", "") },
+		},
+	}
+
+	for _, test := range invalidCases {
+		t.Run(test.name, func(t *testing.T) {
+			if _, err := test.constructor(); err == nil {
+				t.Error("Expected constructor to fail with error, but no error was returned")
+			}
+		})
+	}
+}
+
+func TestStartStopWatch(t *testing.T) {
+	o := &ObserverFuncs{
+		AddFunc: func(s *corev1.Secret) {},
+	}
+
+	for _, test := range defaultWatchersToTest {
+		t.Run(test.name, func(t *testing.T) {
+			testSetup()
+
+			tObs := newTestObserver(o)
+			watcher := test.constructor(t, tObs)
+			// Test weird patterns of stop/start
+			watcher.StopWatch()
+			watcher.StartWatch()
+			watcher.StartWatch()
+			watcher.StopWatch()
+			watcher.StopWatch()
+
+			watcher.StartWatch()
+			defer watcher.StopWatch()
+
+			kubeclientForTest.CoreV1().Secrets(dsft.namespace).Create(&dsft.secret)
+			waitForCondition(t, func() bool {
+				return tObs.CheckOnAddCalled()
+			}, 3)
+
+			testCleanup()
+		})
+	}
+}
+
+func TestNilObserverFuncs(t *testing.T) {
+	testSetup()
+	defer testSetup()
+
+	// Setup two watchers that watch the same Secret, but only have a subset of callbacks
+	o := &ObserverFuncs{
+		AddFunc:    nil,
+		UpdateFunc: nil,
+		DeleteFunc: nil,
+	}
+
+	// The testObserver wrapper struct will invoke the funcs from "o" and track whether they were called
+	tObs := newTestObserver(o)
+	watcher := defaultGlobalSecretWatcherConstructor(t, tObs)
+
+	watcher.StartWatch()
+	defer watcher.StopWatch()
+
+	kubeclientForTest.CoreV1().Secrets(dsft.namespace).Create(&dsft.secret)
+	waitForCondition(t, func() bool {
+		return tObs.CheckOnAddCalled()
```

Comment:
I don't think it make sense to test it has been called as you just set `called` to true. I would just do the add/update/delete operations and make sure they don't introduce any error when the callback funcs are nil.

Reply:
Hmm good catch... I cannot call add/update/delete on a SecretWatcher directly as that is not part of the interface. I will refactor testObs to not consider nil funcs to be "called" by default.